### ISSUE TITLE
pytorch log level API and env var

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#94788]
### 🚀 The feature, motivation and pitch

I'm requesting to please implement an API to allow users to set logging level across pytorch 

For example here is how I set a consistent log level in my applications across multiple packages on a multi-gpu training setup:
```
    log_level = logging.WARNING if dist.get_rank() == 0 else logging.ERROR
    
    datasets.utils.logging.set_verbosity(log_level)
    transformers.utils.logging.set_verbosity(log_level)
    
    from deepspeed.utils import logger as ds_logger
    ds_logger.setLevel(log_level)
```

Would it be possible to be able to do:
```
    torch.utils.logging.set_verbosity(log_level)
```

Each of the above examples supports `DATASETS_VERBOSITY`, `TRANSFORMERS_VERBOSITY` env vars to accomplish the same.

`torch.distributed` uses `LOGLEVEL`, which already helps, but unfortunately doesn't use a domain specific prefix. Ideally it would be some sort of `PYTORCH_LOGLEVEL` so that it won't potentially clash with other packages who too didn't think of using a prefixed env var with a unique package-related prefix.

So to summarize:

1. `torch.utils.logging.set_verbosity(log_level)`
2. `PYTORCH_LOGLEVEL` as another way of setting the log level.

I'm not attached to naming, these are just examples.

wrt implementation you are more than welcome to adapt our feature-full [logging.py](https://github.com/huggingface/transformers/blob/main/src/transformers/utils/logging.py) module. All you need to do is `s/transformers/pytorch/g` and `s/TRANSFORMERS/PYTORCH/g`. All the work is already done.

Thank you.

cc @ezyang @gchanan @zou3519

**Comment by cpuhrsch** [comment]
Marked as triage review, because we should discuss this request. Overly verbose logging has been marked as an UX issue by other users too. I could see this become a wider issue if we don't address it soon and create some centralized logging utilities.

**Comment by ezyang** [comment]
cc @mlazos @Chillee

**Comment by ezyang** [comment]
restoring high priority tag, we plan to do something like this in the near future

**Comment by mlazos** [comment]
I have a [PR](https://github.com/pytorch/pytorch/pull/94858) for a system I'm actively working on for PT 2.0, if we can agree on something there, we could possibly expand it to other PyTorch components.

**Comment by stas00** [comment]
@mlazos, it's your way please make sure there is actual API to use besides env vars.

env vars are fantastic for developers of the component, but when a 3rd party application/framework needs to control it, env vars become very difficult to use and you have to write API anyway to support those env vars overrides, so giving those to the users would make their code more robust, IMHO.

**Comment by albanD** [comment]
cc @kurtamohler

**Comment by vgoklani** [comment]
In the current implementation, the PyTorch compiler is flooding my system logger with DEBUG messages. How do I control the logging level of the compiler? I only want WARNINGS/ERROR messages. thanks!

Please see here for details: https://discuss.pytorch.org/t/the-pytorch-compiler-hijacks-my-logger-and-floods-it-with-debug-messages/173208/2

**Comment by mlazos** [comment]
This landed for torch.compile in https://github.com/pytorch/pytorch/pull/94858

Other components will need to be onboarded to the new system to enable a pytorch-wide system.

**Comment by kurtamohler** [comment]
I would like to help out with expanding the new system to other parts of pytorch. I will probably first focus on expanding it to C++. I plan to think about some of the design for that and then post about it for discussion

**Comment by kurtamohler** [comment]
I'm looking into adding a C++ logging system similar to the one added by https://github.com/pytorch/pytorch/issues/94788. Some of the toughts I've had are:

* Some of the Python-defined log component names and artifact types available in Python will to also need to be available in C++. For instance, we'll need to do this if some parts of a component are defined in Python and other parts in C++

* In order to suppress or enable a particular log from being emitted, the user should not need to know whether it was emitted from C++ or Python, that detail should be opaque. So if a component's log level is changed in Python, that change should also propagate to C++, and vice versa. (Would there ever be a reason not to keep them in sync?)

* The C++ APIs for changing the log state and for adding new registrations internally should be consistent with the Python APIs

* We may want to consider moving `LogRegistry` and `LogState` out of Python and into C++ with Python bindings, so that there's just one shared state and registry for both C++ and Python. There may be some tradeoffs to consider here, but if we do always want to keep the C++ and Python log states in sync, I feel like having just one C++ implementation would be simpler than having two separate Python and C++ implementations that we have to explicitly synchronize

Related to that effort, I think the current Python API for toggling the log state could be improved, and I'd like to discuss some ideas.

`torch._logging.set_logs()` (in [`torch/_logging/_internal.py`](https://github.com/pytorch/pytorch/blob/803a1a041a70fb2cb4b2654710372384df7d7c21/torch/_logging/_internal.py#L119-L140)) is currently the method used to configure log levels for different components and also to toggle specific kinds of log messages. Here's the current function signature:

```python
def set_logs(
    dynamo=DEFAULT_LOG_LEVEL,
    aot=DEFAULT_LOG_LEVEL,
    inductor=DEFAULT_LOG_LEVEL,
    bytecode=False,
    aot_graphs=False,
    aot_joint_graph=False,
    graph=False,
    graph_code=False,
    guards=False,
    recompiles=False,
    output_code=False,
    schedule=False,
):
```

The `dynamo`, `aot`, and `inductor` args are the current components that can be configured with different log levels. In the comments and variable names, these arguments are refered to as "log" arguments, but I think we should instead call them "component" arguments, since they specifically correspond with a component's name. These arguments are of type `int`, and they are expected to correspond with the log levels provided by the `logging` Python module: https://docs.python.org/3/library/logging.html#logging-levels

The boolean args, like `bytecode`, `aot_graphs`, etc can be used to toggle different "artifact" types. An artifact is a specific kind of log, not necessarily tied to a particular component, though it could be. For instance, under the `aot` component, we can toggle the artifacts `aot_graphs` and `aot_joint_graph`. But in general, multiple different components could emit the same kinds of artifacts. `component1` and `component2` could both emit logs of `artifact1`, and turning off `artifact1` would prevent both components from emitting those logs, even if both of the components themselves are still enabled.

I think the current interface for `set_logs` is problematic because as we add more components and artifact types, we would continue to add dozens or potentially hundreds more arguments. I think this complicates the interface for users, and it also adds an extra place that developers will need to modify when adding more components and artifacts. And considering consistency between C++ and Python APIs, that would be an additional place we would need to modify.

It think it's also somewhat confusing to have both components and artifacts in a long list like this without an explicit indication of which ones are components and which ones are artifacts. For instance, if something like `torch.logging.set_logs(dynamo=<value>, graph=<value>)` appears in user code (once we make this interface public), it's not explicitly clear that `dynamo` is a component and `graph` is an artifact type.

So I propose changing the interface such that the individual components and artifact types are not explicit arguments in the function signature, and so that components and artifacts are clearly separated. I think there are at least two possible ways to do this:

  1. Change the signature to `set_logs(components: dict[str, int], artifacts: list[str, bool])`

  2. Separate `set_logs` into `set_components(**components: dict[str, int])` and `set_artifacts(**artifacts: dict[str, bool])`

I like option 2 because it still gives us the convenience of using keyword arguments instead of having to explicitly create dictionaries, like we would have to do for option 1. It would be possible to actually do both of these, but having two different ways to do the same thing might not be a good idea.

As for the environment variable `TORCH_LOGS`, we could also consider splitting it up into `TORCH_LOG_COMPONENTS` and `TORCH_LOG_ARTIFACTS`.

I think we should also have some function or functions that list out all the existing component names and artifact types, potentially including a brief description for them. The description could be added to the function that adds the component or artifact to the log registry. It may also be nice to be able to automatically generate sphinx docs for the component names and artifacts that get added to the registry, though I'm not sure if that can be done statically or not

I would be interested in making these changes if we agree to do this or something similar

**Comment by lezcano** [comment]
About the `set_logs` proposal, do you want these functions to take a variadic number of kwargs?
If this is the case, then, inside the function somewhere, you'll still need to check whether the given inputs make sense, so you'll still need to have that long list somewhere.
As such, you might as well have it in the function signature and ask Python to do this check for you. 
This is also more in the intersection of Python and C++ and not simply for Python.
Also, in Python, you can get the dictionary of args that you want via by calling `locals()` at the top of the function.

Now, both semantically and looking at the implementation of set_logs, I think it makes complete sense to split it into two functions, one for components and another one for artefacts.

**Comment by kurtamohler** [comment]
> About the `set_logs` proposal, do you want these functions to take a variadic number of kwargs?

Yes, that is what I was proposing

> If this is the case, then, inside the function somewhere, you'll still need to check whether the given inputs make sense, so you'll still need to have that long list somewhere. As such, you might as well have it in the function signature and ask Python to do this check for you. This is also more in the intersection of Python and C++ and not simply for Python.

That's a very good point. Since it avoids needing to check each arg against a list of existing components/artifacts, explicit arguments seem like the way to go to me now

Another point I just realized is that having explicit arguments gives us a simple way to document what each one does--just add it to the docstring, rather than having to make sphinx iterate through a list of components/artifacts

**Comment by ezyang** [comment]
> We may want to consider moving LogRegistry and LogState out of Python and into C++ with Python bindings, so that there's just one shared state and registry for both C++ and Python. There may be some tradeoffs to consider here, but if we do always want to keep the C++ and Python log states in sync, I feel like having just one C++ implementation would be simpler than having two separate Python and C++ implementations that we have to explicitly synchronize

Python logging architecture is fairly complicated and I don't really want to replicate it in C++. It seems better for C++ logs to just get piped to Python, and then have standard Python handling. Some thought will be needed about how to do this efficiently (i.e., without eagerly formatting tons of strings.)

**Comment by mlazos** [comment]
> Separating artifacts and components into their own functions would create a semantic separation between two already separate concepts. As for discoverability, we could put a link to the other function in the doc strings. I don't want to speak for Stas, but reading the discussion, it seemed like he thought that putting the artifacts and components in the same function is bit confusing. I was a bit confused by it as well. But it's probably not a huge difference, so either way is alright with me
> 
> I can start documenting the user-facing API. Would it be alright if I generate sphinx docs for a non-public API, and mark it as "prototype"? I guess it's probably possible to disable it from being generated by default and turn it on with some setting

Yes I'd still prefer to keep the single API. I think your plan for the docs sounds good - not sure if it matters to disable the generation by default, up to you.

**Comment by kurtamohler** [comment]
@mlazos , while writing the documentation, I realized that `torch._logging.set_logs` resets every setting when it's called. If one of the kwargs is not specified, it gets set to the default.

For instance:

```python
set_logs(aot=logging.DEBUG)
set_logs(dynamo=logging.DEBUG)
```

Here, the second call will reset "aot" to its default value of `logging.WARN`. That's probably ok as long as the docs make that clear. But I think we should consider changing it so that `set_logs` only changes the settings that were given to it and does not set the others back to the default. This allows users to split up the `set_logs` calls if they want or need to

**Comment by mlazos** [comment]
@kurtamohler yeah I favored making the state as clear as possible after a call, rather than having to track in your code where the state has changed. All of the internal and OSS cases I've seen just call set_logs once at the beginning of their program anyway. I'm not hard set on this though, I could see it only updating values that have changed and adding a reset() or something.

**Comment by kurtamohler** [comment]
I can see the benefit of resetting to defaults. Probably best to leave it as it is for now

**Comment by kurtamohler** [comment]
Hey @mlazos, I'm experimenting with emitting logging artifacts, and I'd like to share some thoughts based on this example script:

<details><summary>Click to expand</summary>

```python
import torch
import logging

inductor_schedule_log = torch._logging.getArtifactLogger('torch._inductor', 'schedule')

def example_inductor_func():
    if inductor_schedule_log.isEnabledFor(logging.DEBUG):
        inductor_schedule_log.debug('message')

print('----------------------')
torch._logging.set_logs(inductor=logging.DEBUG, schedule=False)
example_inductor_func()

print('----------------------')
torch._logging.set_logs(inductor=logging.DEBUG, schedule=True)
example_inductor_func()

print('----------------------')
torch._logging.set_logs(inductor=logging.CRITICAL, schedule=True)
example_inductor_func()


aot_schedule_log = torch._logging.getArtifactLogger('torch._functorch.aot_autograd', 'schedule')

def example_aot_func():
    if aot_schedule_log.isEnabledFor(logging.DEBUG):
        aot_schedule_log.debug('message')

print('----------------------')
torch._logging.set_logs(aot=logging.DEBUG, schedule=False)
example_aot_func()

print('----------------------')
torch._logging.set_logs(aot=logging.DEBUG, schedule=True)
example_aot_func()

print('----------------------')
torch._logging.set_logs(aot=logging.CRITICAL, schedule=True)
example_aot_func()
```

The output is:

```
----------------------
----------------------
[2023-04-18 14:53:18,387] torch._inductor.__schedule: [DEBUG] message
----------------------
[2023-04-18 14:53:18,387] torch._inductor.__schedule: [DEBUG] message
----------------------
----------------------
[2023-04-18 14:53:18,387] torch._functorch.aot_autograd.__schedule: [DEBUG] message
----------------------
[2023-04-18 14:53:18,387] torch._functorch.aot_autograd.__schedule: [DEBUG] message
```

</details>

First, I'd like to double check that I'm doing this in the way you intended. I'm trying to mimic what I see in the codebase.

I'd like to suggest changing the output of `torch._logging.getArtifactLogger` to be an object that can automatically check whether the artifact is enabled for a particular log level when `.debug(...)`, `.warning(...)`, etc are called on it. That way we wouldn't have to wrap every `<artifact>_log.debug(...)` call with a conditional `<artifact>_log.isEnabledFor(logging.DEBUG)` check, and so on.

You mentioned before (in https://github.com/pytorch/pytorch/issues/94788#issuecomment-1503985886) that artifacts are associated with a specific component. But in the above example, it seems that the "schedule" artifact works the same way under "inductor" and "aot" components. Perhaps that's because of an update you've made recently, I'm not sure.

Finally, I'm not sure if `isEnabledFor()` is working as expected. When I set `inductor=logging.CRITICAL`, the inductor schedule debug message still gets emitted, even though debug is less than critical. Same for aot. I thought that the component's log level is supposed to control whether an artifact under that component will be emitted or not. Or are artifacts always supposed to be emitted if enabled, no matter what the component's log level is? If this is the intended behavior, should we consider changing it?

**Comment by mlazos** [comment]
> 

I need to update the design doc on this to make this more clear. So currently it is an OR relationship. If the component or the artifact are enabled, you will see the artifact. This is by design. In the python logging documentation it mentions this explicitly [here](https://docs.python.org/3/library/logging.html#:~:text=If%20this%20attribute,of%20ancestor%20loggers.). Basically, if a child logger accepts a message, it will pass the message to all handlers of its ancestor loggers without considering the level of ancestor loggers. The reason for this is that you only need to setup handlers at the root of your hierarchy rather than iterating over the tree of all loggers. The reason for this is that if you want to see all logs for a component, I assume that you would also like to see the artifacts of that component. This so far has held true.

**Comment by mlazos** [comment]
> > So currently it is an OR relationship. If the component or the artifact are enabled, you will see the artifact.
> 
> In the example I shared in my last comment, the first case sets `inductor=logging.DEBUG, schedule=False`, and it then generates a debug-level "schedule" artifact under the "inductor" component. The "component is enabled at debug level" is True and "artifact is enabled" is False, so "component is enabled at debug level" OR "artifact is enabled" should be True. So if I understand what you said correctly, this should emit the artifact, but it's actually getting suppressed. Am I missing something, or is the behavior unexpected?
> 
> Also, as my example shows, we seem to be completely free to create multiple separate `Logger` instances that emit the same type of artifact but are assigned to different parent components. We can control both of those logger instances with the same `torch._logging.set_logs(<artifact name>=...)` setting. It seems like we currently do have this ability. Am I misunderstanding something? (To clarify something that I might have accidentally implied, I don't mean that we should be able to emit individual log messages of a particular artifact type without a parent component. I agree that all log messages should be emitted under a specific component.)
> 
> I'm still curious what you think about removing the need for calling `isEnabledFor` before emitting an artifact. I can make that change if you'd like

Yeah one other piece is that some artifacts are "off_by_default" (should probably rename this to explicit) basically if I thought an artifact is too spammy, you should be able to only enable it if explicitly named. You happened to pick the single artifact where this is set to true lol.

So `isEnabledFor` is not necessary, but I put it on artifacts that are expensive to generate. I don't use it on all artifacts, mainly graphs and schedules because those can get really large. This isn't needed in the default case. An alternative is what Ed added: [LazyStrings](https://github.com/pytorch/pytorch/blob/8a89eec2f8eab23057d2481243c3b2a77b5cbdcb/torch/_dynamo/utils.py#L1437) only get rendered if they are accessed.

**Comment by lezcano** [comment]
As for the or relationship, that's what I assumed from the beginning, but I was surprised to see that `output_code` would not be generated when using the `inductor` tag. See https://github.com/pytorch/pytorch/issues/99144. Which component is the `output_code` artifact associated to?

**Comment by mlazos** [comment]
> As for the or relationship, that's what I assumed from the beginning, but I was surprised to see that `output_code` would not be generated when using the `inductor` tag. See #99144. Which component is the `output_code` artifact associated to?

It's associated with Inductor, that's also `off_by_default` oops

**Comment by mlazos** [comment]
Closing, as the `TORCH_LOGS` work has been completed and is used widely, we should possibly create a separate issue for C++ integration if that is still getting implemented.
